hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/BucketIndexSupport.scala

```diff
@@ -0,0 +1,194 @@
+import org.apache.spark.util.collection.BitSet
+
+import scala.collection.{JavaConverters, mutable}
+
+class BucketIndexSupport(metadataConfig: HoodieMetadataConfig, schema: StructType) {
+
+  private val log = LogManager.getLogger(getClass);
+
+  private val keyGenerator =
+    HoodieSparkKeyGeneratorFactory.createKeyGenerator(metadataConfig.getProps)
+
+  private lazy val avroSchema = AvroConversionUtils.convertStructTypeToAvroSchema(schema, "record", "")
+
+  /**
+   * the configured bucket field for the table
+   * table may not set bucket fields, it will use record key fields
+   */
+  private val indexBucketHashFieldsOpt: Option[java.util.List[String]] = {
+    val bucketHashFields = metadataConfig.getStringOrDefault(HoodieIndexConfig.BUCKET_INDEX_HASH_FIELD,
+      metadataConfig.getString(HoodieTableConfig.RECORDKEY_FIELDS))
+    if (bucketHashFields == null || bucketHashFields.isEmpty) {
+      Option.apply(null)
+    } else {
+      Option.apply(JavaConverters.seqAsJavaListConverter(bucketHashFields.split(",")).asJava)
+    }
+  }
+
+  def getCandidateFiles(allFiles: Seq[FileStatus], bucketIds: BitSet): Set[String] = {
+    val candidateFiles: mutable.Set[String] = mutable.Set.empty
+    for (file <- allFiles) {
+      val fileId = FSUtils.getFileIdFromFilePath(file.getPath)
+      val fileBucketId = BucketIdentifier.bucketIdFromFileId(fileId)
+      if (bucketIds.get(fileBucketId)) {
+        candidateFiles += file.getPath.getName
+      }
+    }
+    candidateFiles.toSet
+  }
+
+  def filterQueriesWithBucketHashField(queryFilters: Seq[Expression]): Option[BitSet] = {
+    val bucketNumber = metadataConfig.getInt(HoodieIndexConfig.BUCKET_INDEX_NUM_BUCKETS)
+    if (indexBucketHashFieldsOpt.isEmpty || queryFilters.isEmpty) {
+      None
+    } else {
+      var matchedBuckets: BitSet = null
+      if (indexBucketHashFieldsOpt.get.size == 1) {
+        matchedBuckets = getBucketsBySingleHashFields(queryFilters.reduce(And), indexBucketHashFieldsOpt.get.get(0), bucketNumber)
+      } else {
+        matchedBuckets = getBucketsByMultipleHashFields(queryFilters,
+          JavaConverters.asScalaBufferConverter(indexBucketHashFieldsOpt.get).asScala.toSet, bucketNumber)
+      }
+
+      val numBucketsSelected = matchedBuckets.cardinality()
+
+      // None means all the buckets need to be scanned
+      if (numBucketsSelected == bucketNumber) {
+        log.info("the query predicates does not specify equality for all the hashing fields, fallback other index")
+        None
+      } else {
+        Some(matchedBuckets)
+      }
+    }
+  }
+
+  // multiple hash fields only support Equality expression by And
+  private def getBucketsByMultipleHashFields(queryFilters: Seq[Expression], indexBucketHashFields: Set[String], numBuckets: Int): BitSet = {
+    val hashValuePairs = queryFilters.map(expr => getEqualityFieldPair(expr, indexBucketHashFields)).filter(pair => pair != null)
```

Comment:
Do we validate the formation of the `queryFilters`, should be conjunction of multiple equation bucket key filters.

Reply:
yes, expression will be split use PredicateHelper.splitConjunctivePredicates